firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,544 @@
+            assert patch0_URL != patch1_URL,\
+                'Partial and Complete download URLs are different'
+
+        return self.active_update.selected_patch_type == 'complete'
+
+    @property
+    def os_version(self):
+        """Returns information about the OS version
+
+        :returns: the OS version
+        """
+
+        return self.marionette.execute_script("""
+            Cu.import("resource://gre/modules/Services.jsm");
+
+            let osVersion;
+            try {
+              osVersion = Services.sysinfo.getProperty("name") + " " +
+                          Services.sysinfo.getProperty("version");
+            }
+            catch (ex) {
+            }
+
+            if (osVersion) {
+              try {
+                osVersion += " (" + Services.sysinfo.getProperty("secondaryLibrary") + ")";
+              }
+              catch (e) {
+                // Not all platforms have a secondary widget library,
+                // so an error is nothing to worry about.
+              }
+              osVersion = encodeURIComponent(osVersion);
+            }
+            return osVersion;
+        """)
+
+    @property
+    def patch_info(self):
+        """ Returns information of the active update in the queue."""
+        info = {'channel': self.update_channel.channel}
+
+        if (self.active_update.exists):
+            info['buildid'] = self.active_update.buildID
+            info['is_complete'] = self.is_complete_update
+            info['size'] = self.active_update.selected_patch_size
+            info['type'] = self.update_type
+            info['url_mirror'] = \
+                self.active_update.selected_patch_final_url or 'n/a'
+            info['version'] = self.active_update.version
+
+        return info
+
+    @property
+    def staging_directory(self):
+        return self.marionette.execute_script("""
+            let aus = Cc['@mozilla.org/updates/update-service;1']
+                .getService(Ci.nsIApplicationUpdateService);
+            var dir = aus.getUpdatesDirectory();
+            return dir.path;
+        """)
+
+    @property
+    def update_type(self):
+        """Returns the type of the active update."""
+        return self.active_update.type
+
+    def update_url(self, force=False):
+        """Retrieve the AUS update URL the update snippet is retrieved from
+
+        :param force: Boolean flag to force an update check
+
+        :returns: The URL of the update snippet
+        """
+
+        url = self.prefs.get_pref(self.PREF_APP_UPDATE_URL, '')
+        # get the next two prefs from the default branch
+        dist = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION, True) or 'default'
+        dist_version = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION_VERSION, True) or 'default'
+
+        if not url:
+            return None
+
+        # Not all placeholders are getting replaced correctly by formatURL
+        url = re.sub(r'%PRODUCT%', self.app_info.name, url)
```

Comment:
I would suggest we keep using the string replace method here, which is much faster than any regex.